src/VisualStudio/Core/Def/Workspace/VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory.cs

```diff
@@ -0,0 +1,105 @@
+            // We will record all generators for the main workspace in one bucket, and any other generators running in other
+            // workspaces (interactive, for example) will be put in a different bucket. This allows us to report the telemetry
+            // from the primary workspace on solution closed, while not letting the unrelated runs pollute those numbers.
+            if (workspaceServices.Workspace is VisualStudioWorkspace)
+            {
+                EnsureSubscribedToSolutionEvents();
+                return _visualStudioWorkspaceInstance;
+            }
+            else
+            {
+                return _otherWorkspacesInstance;
+            }
+        }
+
+        private void EnsureSubscribedToSolutionEvents()
+        {
+            if (Interlocked.CompareExchange(ref _subscribedToSolutionEvents, 1, 0) == 0)
+            {
+                Task.Run(async () =>
+                {
+                    var shellService = await _serviceProvider.GetServiceAsync<SVsSolution, IVsSolution>(_threadingContext.JoinableTaskFactory).ConfigureAwait(true);
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(_threadingContext.DisposalToken);
+                    shellService.AdviseSolutionEvents(this, out _);
+                }, _threadingContext.DisposalToken);
+            }
+        }
+
+        public void ReportOtherWorkspaceTelemetry()
+        {
+            _otherWorkspacesInstance.ReportStatisticsAndClear(FunctionId.SourceGenerator_OtherWorkspaceSessionStatistics);
+        }
+
+        int IVsSolutionEvents.OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnQueryCloseProject(IVsHierarchy pHierarchy, int fRemoving, ref int pfCancel) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnBeforeCloseProject(IVsHierarchy pHierarchy, int fRemoved) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnAfterLoadProject(IVsHierarchy pStubHierarchy, IVsHierarchy pRealHierarchy) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnQueryUnloadProject(IVsHierarchy pRealHierarchy, ref int pfCancel) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnBeforeUnloadProject(IVsHierarchy pRealHierarchy, IVsHierarchy pStubHierarchy) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnAfterOpenSolution(object pUnkReserved, int fNewSolution) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnQueryCloseSolution(object pUnkReserved, ref int pfCancel) => VSConstants.E_NOTIMPL;
+
+        int IVsSolutionEvents.OnBeforeCloseSolution(object pUnkReserved)
+        {
+            // Report the telemetry now before the solution is closed; since this will be reported per solution session ID, it means
+            // we can distinguish how many solutions have generators versus just overall sessions.
+            _visualStudioWorkspaceInstance.ReportStatisticsAndClear(FunctionId.SourceGenerator_SolutionStatistics);
```

Comment:
worried about ths perf impact.

Reply:
Ah, I see GitHub somehow ate my prior comment. Any idea if there's a way I can capture the context but still report telemetry async? Because that'd be perfect here.

Reply:
Chatting with the team it sounds like what I want doesn't immediately exist; I'll do a validation run with this to ensure we don't see regressions.

Reply:
ok.  i look forward to DaveKean's feedback here :D